### ISSUE TITLE
False positive unused_attributes lint on doc(hidden) attribute on associated type

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust#96890]
The warning added by #96008 claims that `doc(hidden)` on an impl item is ignored by rustdoc, but it isn't necessarily.

The warning says:

> whether the impl item is \`doc(hidden)\` or not entirely depends on the corresponding trait item

which is not true in the following repro:

```rust
pub trait Trait {
    type A;
    type B;
}

pub struct Struct;

impl Trait for Struct {
    type A = [(); "interesting type".len()];
    #[doc(hidden)]
    type B = [(); "nasty uninteresting type".len()];
}
```

```console
$ cargo check
warning: `#[doc(hidden)]` is ignored on trait impl items
  --> src/lib.rs:10:5
   |
10 |     #[doc(hidden)]
   |     ^^^^^^^^^^^^^^ help: remove this attribute
   |
   = note: `#[warn(unused_attributes)]` on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: whether the impl item is `doc(hidden)` or not entirely depends on the corresponding trait item
```

This warning is a false positive. Rustdoc **does** pay attention to this attribute. For the code above, the rendered documentation of `Struct` and `Trait` looks like:

<table><tr><td><img src="https://user-images.githubusercontent.com/1940490/167544111-4e7319ee-4c83-4b04-b459-62b3fe9912da.png" height="185"></td><td><img src="https://user-images.githubusercontent.com/1940490/167544007-63ca4464-7230-4924-b33d-2814daa37a42.png" height="300"></td></tr></table>

After removing the supposedly unused `doc(hidden)` attribute and rendering again, the "nasty uninteresting type" that we wanted not shown is going to get rendered on both pages:

<table><tr><td><img src="https://user-images.githubusercontent.com/1940490/167544117-58f15e02-ef35-4c1e-9c97-f426ec1a33ea.png" height="185"></td><td><img src="https://user-images.githubusercontent.com/1940490/167544122-a62e3ab8-2ce8-42eb-99b0-9627dd97eda6.png" height="300"></td></tr></table>

(@fmease @lcnr)

### Meta

`rustc --version --verbose`:
```
rustc 1.62.0-nightly (88860d547 2022-05-09)
binary: rustc
commit-hash: 88860d5474a32f507dde8fba8df35fd2064f11b9
commit-date: 2022-05-09
host: x86_64-unknown-linux-gnu
release: 1.62.0-nightly
LLVM version: 14.0.1
```

**Comment by fmease** [comment]
My bad, I did not notice that _at all_ 😳. I can adjust the lint once it's decided how we are going to proceed with my changes in general (see / blocked on https://github.com/rust-lang/rust/pull/96008#issuecomment-1122191794).

@rustbot claim

**Comment by fmease** [comment]
In summary, `#[doc(hidden)]` on _methods_ in trait impl blocks does not lead to any visual changes in the documentation whereas on associated types and constants, it hides the implementation (or “the thing to the right of `=`”).

This is weirdly deviates from my (former) personal mental model of how `#[doc(hidden)]` is supposed to work and used. As far as I know, in all other cases, it leads to items, impl blocks, fields (“names” or “declarations”) not being shown. Here, however, the “name” keeps being shown, only the “value” is hidden.
Out of curiosity, @ dtolnay, how / why do you use this feature? Do you use it to reduce visual clutter?

**Comment by fmease** [comment]
[Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/Validy.20checks.20for.20.60.23.5Bdoc.28hidden.29.5D.60/near/281846219)

**Comment by fmease** [comment]
Should I remove the lint entirely or should I adjust `#[doc(hidden)]`?

**Comment by SergioBenitez** [comment]
Just wanted to add that this is also affecting `rocket`, but in a different way:

```rust
impl<'a> Origin<'a> {
    /// The root: `'/'`.
    #[doc(hidden)]
    pub const ROOT: Origin<'static> = Origin::const_new("/", None);
}
```

<img width="500" src="https://user-images.githubusercontent.com/1480321/169919843-ea08a980-9178-42a2-9cb0-c306a758edb9.png">

The `const` isn't there at all without the `doc(hidden)`:

<img width="500" src="https://user-images.githubusercontent.com/1480321/169919868-67957937-2246-4da2-9936-701ac2ef9896.png">


> In summary, `#[doc(hidden)]` on _methods_ in trait impl blocks does not lead to any visual changes in the documentation whereas on associated types and constants, it hides the implementation (or “the thing to the right of the `=`”).

Based on the above, this isn't correct: it actually hides the entire thing.

**Comment by fmease** [comment]
In your case, it's inherent associated items being accidentally flagged which is #97205 which was closed by #97208 meaning the most recent nightly should not have your issue anymore.
